File: app/src/main/java/com/veyndan/paper/reddit/ui/recyclerview/itemdecoration/TreeInsetItemDecoration.java
package com.veyndan.paper.reddit.ui.recyclerview.itemdecoration;

import android.content.Context;
import android.graphics.Rect;
import android.support.annotation.DimenRes;
import android.support.annotation.NonNull;
import android.support.annotation.Px;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.hannesdorfmann.adapterdelegates3.ListDelegationAdapter;
import com.veyndan.paper.reddit.util.Node;

import java.util.List;

import static com.google.common.base.Preconditions.checkArgument;

public class TreeInsetItemDecoration extends RecyclerView.ItemDecoration {

    @Px private final int childInsetMultiplier;

    public TreeInsetItemDecoration(@NonNull final Context context,
                                   @DimenRes final int childInsetMultiplierRes) {
        childInsetMultiplier = context.getResources().getDimensionPixelOffset(childInsetMultiplierRes);
    }

    @Override
    public void getItemOffsets(final Rect outRect, final View view, final RecyclerView parent,
                               final RecyclerView.State state) {
        checkArgument(parent.getAdapter() instanceof ListDelegationAdapter,
                "RecyclerView's Adapter must implement ListDelegationAdapter<List<Node<?>>> in " +
                        "order for TreeInsetItemDecoration to be used as a decoration");

        final ListDelegationAdapter<List<Node<?>>> listDelegationAdapter =
                (ListDelegationAdapter<List<Node<?>>>) parent.getAdapter();
        final int position = parent.getChildAdapterPosition(view);
        final List<Node<?>> nodes = listDelegationAdapter.getItems();

        final int inset;
        if (position == RecyclerView.NO_POSITION) {
            inset = 0;
        } else {
            final Node<?> node = nodes.get(position);
            final int depth = node.getDepth();
            inset = depth * childInsetMultiplier;
        }

        outRect.set(inset, 0, 0, 0);
    }
}

File: app/src/main/java/com/veyndan/paper/reddit/ui/recyclerview/itemdecoration/TreeInsetItemDecoration.kt
package com.veyndan.paper.reddit.ui.recyclerview.itemdecoration

import android.content.Context
import android.graphics.Rect
import android.support.annotation.DimenRes
import android.support.annotation.Px
import android.support.v7.widget.RecyclerView
import android.view.View
import com.hannesdorfmann.adapterdelegates3.ListDelegationAdapter
import com.veyndan.paper.reddit.util.Node

class TreeInsetItemDecoration(context: Context,
                              @DimenRes childInsetMultiplierRes: Int) : RecyclerView.ItemDecoration() {

    @Px private val childInsetMultiplier: Int = context.resources.getDimensionPixelOffset(childInsetMultiplierRes)

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView,
                                state: RecyclerView.State?) {
        check(parent.adapter is ListDelegationAdapter<*>) { "RecyclerView's Adapter must implement ListDelegationAdapter<List<Node<?>>> in order for TreeInsetItemDecoration to be used as a decoration" }

        val listDelegationAdapter: ListDelegationAdapter<List<Node<*>>> = parent.adapter as ListDelegationAdapter<List<Node<*>>>
        val position: Int = parent.getChildAdapterPosition(view)
        val nodes: List<Node<*>> = listDelegationAdapter.items

        val inset: Int = if (position == RecyclerView.NO_POSITION) 0 else {
            val depth: Int = nodes[position].depth
            depth * childInsetMultiplier
        }

        outRect.set(inset, 0, 0, 0)
    }
}
